src/DotNetty.Transport/Channels/Groups/DefaultChannelGroupCompletionSource.cs

```diff
@@ -0,0 +1,146 @@
+                    bool success = x.IsCompleted && !x.IsFaulted && !x.IsCanceled;
+                    bool callSetDone;
+                    lock (this)
+                    {
+                        if (success)
+                        {
+                            this.successCount++;
+                        }
+                        else
+                        {
+                            this.failureCount++;
+                        }
+
+                        callSetDone = this.successCount + this.failureCount == this.futures.Count;
+                        Debug.Assert(this.successCount + this.failureCount <= this.futures.Count);
+                    }
+
+                    if (callSetDone)
+                    {
+                        if (this.failureCount > 0)
+                        {
+                            var failed = new List<KeyValuePair<IChannel, Exception>>();
+                            foreach (KeyValuePair<IChannel, Task> ft in this.futures)
+                            {
+                                IChannel c = ft.Key;
+                                Task f = ft.Value;
+                                if (f.IsFaulted || f.IsCanceled)
+                                {
+                                    if (f.Exception != null)
+                                    {
+                                        failed.Add(new KeyValuePair<IChannel, Exception>(c, f.Exception.InnerException));
+                                    }
+                                }
+                            }
+                            this.TrySetException(new ChannelGroupException(failed));
+                        }
+                        else
+                        {
+                            this.TrySetResult(0);
+                        }
+                    }
+                });
+            }
+
+            // Done on arrival?
+            if (futures.Count == 0)
+            {
+                this.TrySetResult(0);
+            }
+        }
+
+        public IChannelGroup Group
+        {
+            get { return this.group; }
+        }
+
+        public Task Find(IChannel channel)
+        {
+            return this.futures[channel];
+        }
+
+        public bool IsPartialSucess()
+        {
+            lock (this)
+            {
+                return this.successCount != 0 && this.successCount != this.futures.Count;
+            }
+        }
+
+        public bool IsSucess()
+        {
+            return this.Task.IsCompleted && !this.Task.IsFaulted && !this.Task.IsCanceled;
+        }
+
+        public bool IsPartialFailure()
+        {
+            lock (this)
+            {
+                return this.failureCount != 0 && this.failureCount != this.futures.Count;
+            }
+        }
+
+        public ChannelGroupException Cause
+        {
+            get { return (ChannelGroupException)this.Task.Exception.InnerException; }
+        }
+
+        public Task Current
```

Comment:
switch to impl `IEnumerable<Task>` as suggested on the intf => you'd only need to impl `.GetEnumerator()` returning `this.futures.Values.GetEnumerator()`